include/solvers/mpm_explicit.tcc

```diff
@@ -117,190 +109,36 @@ bool mpm::MPMExplicit<Tdim>::solve() {
-        std::bind(&mpm::NodeBase<Tdim>::compute_velocity,
-                  std::placeholders::_1),
-        std::bind(&mpm::NodeBase<Tdim>::status, std::placeholders::_1));
-
-    if (interface_) {
-      // Map multimaterial properties from particles to nodes
-      mesh_->iterate_over_particles(std::bind(
-          &mpm::ParticleBase<Tdim>::map_multimaterial_mass_momentum_to_nodes,
-          std::placeholders::_1));
-
-      // Map multimaterial displacements from particles to nodes
-      mesh_->iterate_over_particles(std::bind(
-          &mpm::ParticleBase<Tdim>::map_multimaterial_displacements_to_nodes,
-          std::placeholders::_1));
-
-      // Map multimaterial domain gradients from particles to nodes
-      mesh_->iterate_over_particles(std::bind(
-          &mpm::ParticleBase<Tdim>::map_multimaterial_domain_gradients_to_nodes,
-          std::placeholders::_1));
-
-      // Compute multimaterial change in momentum
-      mesh_->iterate_over_nodes(std::bind(
-          &mpm::NodeBase<Tdim>::compute_multimaterial_change_in_momentum,
-          std::placeholders::_1));
-
-      // Compute multimaterial separation vector
-      mesh_->iterate_over_nodes(std::bind(
-          &mpm::NodeBase<Tdim>::compute_multimaterial_separation_vector,
-          std::placeholders::_1));
-
-      // Compute multimaterial normal unit vector
-      mesh_->iterate_over_nodes(std::bind(
-          &mpm::NodeBase<Tdim>::compute_multimaterial_normal_unit_vector,
-          std::placeholders::_1));
-    }
+    // Map material properties to nodes
+    contact_->compute_contact_forces();
 
     // Update stress first
-    if (this->stress_update_ == mpm::StressUpdate::USF)
-      this->compute_stress_strain(phase);
-
-      // Spawn a task for external force
-#pragma omp parallel sections
-    {
-#pragma omp section
-      {
-        // Iterate over each particle to compute nodal body force
-        mesh_->iterate_over_particles(
-            std::bind(&mpm::ParticleBase<Tdim>::map_body_force,
-                      std::placeholders::_1, this->gravity_));
-
-        // Apply particle traction and map to nodes
-        mesh_->apply_traction_on_particles(this->step_ * this->dt_);
-
-        // Iterate over each node to add concentrated node force to external
-        // force
-        if (set_node_concentrated_force_)
-          mesh_->iterate_over_nodes(std::bind(
-              &mpm::NodeBase<Tdim>::apply_concentrated_force,
-              std::placeholders::_1, phase, (this->step_ * this->dt_)));
-      }
-
-#pragma omp section
-      {
-        // Spawn a task for internal force
-        // Iterate over each particle to compute nodal internal force
-        mesh_->iterate_over_particles(
-            std::bind(&mpm::ParticleBase<Tdim>::map_internal_force,
-                      std::placeholders::_1));
-      }
-    }  // Wait for tasks to finish
+    mpm_scheme_->precompute_stress_strain(phase, pressure_smoothing_);
```

Comment:
Any reason why pressure_smoothing is included in the compute_stress_strain? Isn't supposed to be separate? I think we always need to separate the current stress state and the smoothed one because any plastic constitutive models (especially critical state model) are not happy when the stress state changes.

Reply:
This is just a flag to say you want to pressure smoothing or not